File: LeetCode/Kai/268MissingNumber.java
//solution 1
public class Solution {
    public int missingNumber(int[] nums) {
        int len = nums.length;
        int sum = len*(len+1)/2;
        for(int i=0; i<len; i++) sum -= nums[i];
        return sum;
    }
    
}

//solution 2: bit manipulation
public class Solution {
    public int missingNumber(int[] nums) {
        int base = 0;
        for(int i=0; i<nums.length; i++) {
            base = base ^ i ^ nums[i];   //a ^ b ^ b == a
        }
        return base^nums.length;
    }
}

File: LeetCode/Peng/268MissingNumber.cpp
/**
 * Only computation, a little faster.
 **/ 

class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int n = nums.size();
        if (!n) return 0;
        int total = (1 + n) * n / 2;
        for (int i = 0; i < n; i++) {
            total -= nums.at(i);
        }
        return total;
    }
};

/**
 * Hash version, a little slower.
 **/ 
class Solution {
public:
    int missingNumber(vector<int>& nums) {
        int n = nums.size();
        if (!n) return 0;
        unordered_map<int, int> records;
        for (int i = 0; i < n; i++) 
            records[nums.at(i)] = 1;
        for (int i = 0; i <= n; i++) {
            if (!records.count(i))
                return i;
        }
    }
};